bpf/flows.c

```diff
@@ -65,89 +116,177 @@ static inline int fill_iphdr(struct iphdr *ip, void *data_end, struct flow *flow
+        memset (&(flow_id->src_ip),0, sizeof(struct in6_addr));
+        memset (&(flow_id->dst_ip),0, sizeof(struct in6_addr));
+        flow_id->protocol = 0;
+        flow_id->src_port = 0;
+        flow_id->dst_port = 0;
     }
     return SUBMIT;
 }
 
-// parses flow information for a given direction (ingress/egress)
-static inline int flow_parse(struct __sk_buff *skb, u8 direction) {
+static inline void export_flow_id (flow *my_flow_key, flow_id_v my_flow_id, u8 direction) {
+    my_flow_key->protocol = my_flow_id.eth_protocol;
+    my_flow_key->direction = direction;
+    __builtin_memcpy(my_flow_key->data_link.src_mac, my_flow_id.src_mac, ETH_ALEN);
+    __builtin_memcpy(my_flow_key->data_link.dst_mac, my_flow_id.dst_mac, ETH_ALEN);
+    my_flow_key->network.src_ip = my_flow_id.src_ip;
+    my_flow_key->network.dst_ip = my_flow_id.dst_ip;
+    my_flow_key->transport.src_port = my_flow_id.src_port;
+    my_flow_key->transport.dst_port = my_flow_id.dst_port;
+    my_flow_key->transport.protocol = my_flow_id.protocol;
+}
 
+static inline int flow_monitor (struct __sk_buff *skb, u8 direction) {
     // If sampling is defined, will only parse 1 out of "sampling" flows
     if (sampling != 0 && (bpf_get_prandom_u32() % sampling) != 0) {
         return TC_ACT_OK;
     }
-
-    void *data = (void *)(long)skb->data;
     void *data_end = (void *)(long)skb->data_end;
+    void *data = (void *)(long)skb->data;
+    flow_id_v my_flow_id;
+    int rc = TC_ACT_OK;
+    u32 flags = 0;
+    flow_record *flow_event;
+    flow_metrics *my_flow_counters;
+    __u64 current_time = bpf_ktime_get_ns();
 
-    struct flow *flow = bpf_ringbuf_reserve(&flows, sizeof(struct flow), 0);
-    if (!flow) {
+    struct ethhdr *eth = data;
+    if (fill_ethhdr(eth, data_end, &my_flow_id, &flags) == DISCARD) {
         return TC_ACT_OK;
     }
+    if (direction == INGRESS) {
+        my_flow_counters = bpf_map_lookup_elem(&xflow_metric_map_ingress, &my_flow_id);
+    } else {
+        my_flow_counters = bpf_map_lookup_elem(&xflow_metric_map_egress, &my_flow_id);
+    }
 
-    struct ethhdr *eth = data;
-    if (fill_ethhdr(eth, data_end, flow) == DISCARD) {
-        bpf_ringbuf_discard(flow, 0);
+    if (my_flow_counters != NULL) {
+        my_flow_counters->packets += 1;
+        my_flow_counters->bytes += skb->len;
+        my_flow_counters->last_pkt_ts = current_time;
+        if (flags & TCP_FIN_FLAG || flags & TCP_RST_FLAG) {
+            /* Need to evict the entry and send it via ring buffer */
+            flow_event = bpf_ringbuf_reserve(&flows, sizeof(flow_record), 0);
+            if (!flow_event) {
+                return rc;
+            }
+            export_flow_id(&flow_event->flow_key, my_flow_id, 1);
+            // flow_event->metrics.packets = my_flow_counters->packets;
+            // flow_event->metrics.bytes = my_flow_counters->bytes;
+            // flow_event->metrics.last_pkt_ts = my_flow_counters->last_pkt_ts;
```

Comment:
If we are not going to use this code, let's remove it (we already have the Git history).